x/incentive/keeper/rewards_test.go

```diff
@@ -0,0 +1,341 @@
+	suite.keeper.IterateRewardPeriods(suite.ctx, func(rp types.RewardPeriod) (stop bool) {
+		rewardPeriods = append(rewardPeriods, rp)
+		return false
+	})
+	suite.Equal(2, len(rewardPeriods))
+
+	var claimPeriods types.ClaimPeriods
+	suite.keeper.IterateClaimPeriods(suite.ctx, func(cp types.ClaimPeriod) (stop bool) {
+		claimPeriods = append(claimPeriods, cp)
+		return false
+	})
+	suite.Equal(2, len(claimPeriods))
+
+	var claims types.Claims
+	suite.keeper.IterateClaims(suite.ctx, func(c types.Claim) (stop bool) {
+		claims = append(claims, c)
+		return false
+	})
+	suite.Equal(2, len(claims))
+
+	var genIDs types.GenesisClaimPeriodIDs
+	suite.keeper.IterateClaimPeriodIDKeysAndValues(suite.ctx, func(denom string, id uint64) (stop bool) {
+		genID := types.GenesisClaimPeriodID{Denom: denom, ID: id}
+		genIDs = append(genIDs, genID)
+		return false
+	})
+	suite.Equal(2, len(genIDs))
+}
+
+func (suite *KeeperTestSuite) TestHelperMethods() {
+	rp := types.NewRewardPeriod("bnb", suite.ctx.BlockTime(), suite.ctx.BlockTime().Add(time.Hour*168), c("ukava", 100000000), suite.ctx.BlockTime().Add(time.Hour*168*2), time.Hour*8766)
+	suite.keeper.SetRewardPeriod(suite.ctx, rp)
+	suite.keeper.SetNextClaimPeriodID(suite.ctx, "bnb", 1)
+	suite.NotPanics(func() {
+		suite.keeper.HandleRewardPeriodExpiry(suite.ctx, rp)
+	})
+	_, found := suite.keeper.GetClaimPeriod(suite.ctx, 1, "bnb")
+	suite.True(found)
+
+	addr, _ := sdk.AccAddressFromBech32("kava15qdefkmwswysgg4qxgqpqr35k3m49pkx2jdfnw")
+	c1 := types.NewClaim(addr, c("ukava", 1000000), "bnb", 1)
+	suite.keeper.SetClaim(suite.ctx, c1)
+	suite.NotPanics(func() {
+		suite.keeper.AddToClaim(suite.ctx, addr, "bnb", 1, c("ukava", 1000000))
+	})
+	testC, _ := suite.keeper.GetClaim(suite.ctx, addr, "bnb", 1)
+	suite.Equal(c("ukava", 2000000), testC.Reward)
+
+	suite.NotPanics(func() {
+		suite.keeper.AddToClaim(suite.ctx, addr, "xpr", 1, c("ukava", 1000000))
+	})
+
+	suite.SetupTest()
+	reward := types.NewReward(true, "bnb", c("ukava", 1000000000), time.Hour*7*24, time.Hour*24*365, time.Hour*7*24)
+	suite.NotPanics(func() {
+		suite.keeper.CreateNewRewardPeriod(suite.ctx, reward)
+	})
+	_, found = suite.keeper.GetRewardPeriod(suite.ctx, "bnb")
+	suite.True(found)
+
+}
+
+func (suite *KeeperTestSuite) TestCreateAndDeleteRewardsPeriods() {
+	reward1 := types.NewReward(true, "bnb", c("ukava", 1000000000), time.Hour*7*24, time.Hour*24*365, time.Hour*7*24)
+	reward2 := types.NewReward(false, "xrp", c("ukava", 1000000000), time.Hour*7*24, time.Hour*24*365, time.Hour*7*24)
+	params := types.NewParams(true, types.Rewards{reward1, reward2})
+	suite.keeper.SetParams(suite.ctx, params)
+
+	suite.NotPanics(func() {
+		suite.keeper.CreateAndDeleteRewardPeriods(suite.ctx)
+	})
+
+	_, found := suite.keeper.GetRewardPeriod(suite.ctx, "bnb")
+	suite.True(found)
+	_, found = suite.keeper.GetRewardPeriod(suite.ctx, "xrp")
+	suite.False(found)
+
+}
+
+func (suite *KeeperTestSuite) TestApplyRewardsToCdps() {
+	suite.setupCdpChain()
```

Comment:
`setupCdpChain()` is only called once, but I understand why it's been moved outside of the test since it's a large function. The downside is that it's not very clear what state is - for example, is the expectation of 3 claims in state on line 192 a valid expectation?

Reply:
Not sure there's an easy solution here. What about moving `setupCdpChain()` to keeper_test.go and using it as the default test setup function for all/most methods? Then state expectations will be the same from test to test.

Reply:
I like using default `SetupTest()` with no state, as I think it's easier to reason about most tests if they explicitly add state and don't add state from other modules unless necessary. For big integration tests like this, I'm fine having a separate helper function and just using comments to describe what the expected state is. I agree it's a preference thing and probably doesn't have one 'right' answer.